### DISCUSSION TITLE
tokio::select!   how to handle multiple branch.

### DISCUSSION THREAD

**Discussion by kenmasumitsu** [tokio-rs/tokio#4222] [Q&A] [answered]
Hi, 

tokio::select! need to declare all async operations at the compile time.
If the async operations are not known in compile time, can I use tokio::select! ?

```
   // size of receivers is unknown at the compile time.
    let receivers: Vec<mpsc::Receiver<u32>> = Vec::new();
    receivers.push(rx1);
    receivers.push(rx2);
    ... 
  
    loop {
       tokio::select! {
           // how to monitor all Receivers in receivers ? 
       }
    }
```

I want like [FD_SET](https://linux.die.net/man/3/fd_set) in C.

**Comment by Darksonn** [comment, marked as answer]
If you have a variable number of streams, then you shouldn't be using `select!`. Here are some other options:

 1. Have all the senders send to the same receiver so you only have to receive from one thing.
 2. Use [`tokio_stream::StreamMap`][1].
 3. Use [`futures::stream::SelectAll`][2].

The difference between option 2 and 3 is whether it tells you which receiver the message came from.

[1]: https://docs.rs/tokio-stream/0.1/tokio_stream/struct.StreamMap.html
[2]: https://docs.rs/futures/0.3/futures/stream/struct.SelectAll.html

**Reply by kenmasumitsu** [reply]
Thanks for kindly quick response. I will check the option 2 and 3.

**Reply by Darksonn** [reply]
To use either solution, you are going to need the [`ReceiverStream`][1] wrapper.

[1]: https://docs.rs/tokio-stream/0.1/tokio_stream/wrappers/struct.ReceiverStream.html

**Reply by kenmasumitsu** [reply]
I will check it, too. I really appreciate your support.

**Comment by Noah-Kennedy** [comment]
If the futures all have the same output type, there is the `select_all` function in the `futures` crate: https://docs.rs/futures/0.3.17/futures/future/fn.select_all.html
